momentGW/pbc/rpa.py

```diff
@@ -0,0 +1,498 @@
+
+        kpts = self.kpts
+        Lia = self.integrals.Lia
+
+        self.diag_eri = np.zeros((self.nkpts, self.nkpts), dtype=object)
+        self.d = np.zeros((self.nkpts, self.nkpts), dtype=object)
+
+        self.Liad = np.zeros((self.nkpts, self.nkpts), dtype=object)
+        self.Liadinv = np.zeros((self.nkpts, self.nkpts), dtype=object)
+
+        for q in kpts.loop(1):
+            for kj in kpts.loop(1, mpi=True):
+                kb = kpts.member(kpts.wrap_around(kpts[q] + kpts[kj]))
+                self.diag_eri[q, kb] = np.sum(np.abs(Lia[kj, kb]) ** 2, axis=0) / self.nkpts
+                self.d[q, kb] = util.build_1h1p_energies(
+                    (self.mo_energy_w[kj], self.mo_energy_w[kb]),
+                    (self.mo_occ_w[kj], self.mo_occ_w[kb]),
+                ).ravel()
+
+                self.Liad[q, kb] = Lia[kj, kb] * self.d[q, kb]
+                self.Liadinv[q, kb] = Lia[kj, kb] / self.d[q, kb]
+
+    def integrate(self):
+        """Optimise the quadrature and perform the integration for a
+        given set of k points.
+
+        Returns
+        -------
+        integral : numpy.ndarray
+            Integral array, including the offset part.
+        """
+
+        cput0 = (lib.logger.process_clock(), lib.logger.perf_counter())
+        lib.logger.info(self.gw, "Performing integration")
+        lib.logger.debug(self.gw, "Memory usage: %.2f GB", self._memory_usage())
+
+        # Get the offset integral quadrature
+        quad = self.optimise_offset_quad()
+        cput1 = lib.logger.timer(self.gw, "optimising offset quadrature", *cput0)
+
+        # Perform the offset integral
+        offset = self.eval_offset_integral(quad)
+        cput1 = lib.logger.timer(self.gw, "performing offset integral", *cput1)
+
+        # Get the main integral quadrature
+        quad = self.optimise_main_quad()
+        cput1 = lib.logger.timer(self.gw, "optimising main quadrature", *cput1)
+
+        # Perform the main integral
+        integral = self.eval_main_integral(quad)
+        cput1 = lib.logger.timer(self.gw, "performing main integral", *cput1)
+
+        # Report quadrature error
+        if self.report_quadrature_error:
+            a = 0.0
+            b = 0.0
+            for q in self.kpts.loop(1):
+                for ka in self.kpts.loop(1, mpi=True):
+                    a += np.sum((integral[0, q, ka] - integral[2, q, ka]) ** 2)
+                    b += np.sum((integral[0, q, ka] - integral[1, q, ka]) ** 2)
+            a, b = mpi_helper.allreduce(np.array([a, b]))
+            a, b = a**0.5, b**0.5
+            err = self.estimate_error_clencur(a, b)
+            lib.logger.debug(self.gw, "One-quarter quadrature error: %s", a)
+            lib.logger.debug(self.gw, "One-half quadrature error: %s", b)
+            lib.logger.debug(self.gw, "Error estimate: %s", err)
+
+        return integral[0] + offset
+
+    def build_dd_moments(self, integral=None):
```

Comment:
what's the status of MPI here? there are MPI enabled loops but missing reductions etc.